beam/README.md

```diff
@@ -0,0 +1,169 @@
+
+This step will build a standalone job service jar, and upload it to that path
+for use during cluster creation.
+
+```bash
+./gradlew :beam-runners-flink_2.11-job-server:shadowJar
+```
+
+After the jar has been built, it will need to be uploaded to a GCS path that
+clusters can access during initialization.
+
+```bash
+gsutil cp \
+  ./runners/flink/job-server/build/libs/beam-runners-flink_2.11-job-server-*-SNAPSHOT.jar \
+  <BEAM_JOB_SERVICE_DESTINATION>/beam-runners-flink_2.11-job-server-latest-SNAPSHOT.jar
+```
+
+#### Build the Worker Container Images
+
+This step will build the docker images used by Beam worker tasks.
+
+```bash
+./gradlew docker
+```
+
+This will build a number of docker images with names that look like
+`<USER>-docker-apache.bintray.io/beam/<LANGUAGE>`.  These need to be renamed and
+pushed to a docker repository path that clusters can access during
+initialization.  We recommend tagging these images with the
+`BEAM_SOURCE_VERSION` they were generated from.
+
+```bash
+docker tag \
+  <USER>-docker-apache.bintray.io/beam/<LANGUAGE> \
+  <BEAM_CONTAINER_IMAGE_DESTINATION>/<LANGUAGE>:<BEAM_SOURCE_VERSION>
+docker push <BEAM_CONTAINER_IMAGE_DESTINATION>/<LANGUAGE>:<BEAM_SOURCE_VERSION>
+```
+
+## Creating a Cluster
+
+Running a Beam cluster currently requires the following initialization actions:
+
+  - `docker/docker.sh`
+  - `flink/flink.sh`
+  - `beam/beam_job_service.sh`
+
+In addition, a number of metadata variables need to be set correctly.
+
+| Metadata Key | Value |
+| --- | --- |
+| beam-job-service-snapshot | The GCS path of your JobService jar (see above) |
+| beam-image-enable-pull | `true` |
+| beam-image-repository | `<BEAM_CONTAINER_IMAGE_DESTINATION>` (see above) |
+| beam-image-version | the tag used for container images (default: `latest`) |
+| flink-start-yarn-session | `true` |
+| flink-snapshot-url | URL to a flink 1.5 snapshot such as [this one](https://archive.apache.org/dist/flink/flink-1.5.3/flink-1.5.3-bin-hadoop28-scala_2.11.tgz) |
+
+Beam 2.7 requires Flink 1.5, which is not the default Flink version in Dataproc
+1.2.  Therefore users must specify a custom flink snapshot.
+
+In practice, the best way to handle this many metadata values is through a
+script such as this one:
+
+TODO: anonymize this
+```bash
+CLUSTER_NAME="$1
+INIT_BUCKET="gs://dataproc-initialization-actions"
+INIT_ACTIONS="${INIT_BUCKET}/docker/docker.sh"
+INIT_ACTIONS+=",${INIT_BUCKET}/flink/flink.sh"
+INIT_ACTIONS+=",${INIT_BUCKET}/beam/beam_job_service.sh"
+FLINK_SNAPSHOT="https://archive.apache.org/dist/flink/flink-1.5.3/flink-1.5.3-bin-hadoop28-scala_2.11.tgz"
+METADATA="beam-job-service-snapshot=<...>"
+METADATA+=",beam-image-enable-pull=true"
+METADATA+=",beam-image-repository=<...>"
+METADATA+=",beam-image-version=latest"
+METADATA+=",flink-start-yarn-session=true"
+METADATA+=",flink-snapshot-url=${FLINK_SNAPSHOT}"
+
+gcloud dataproc clusters create "${CLUSTER_NAME}" \
+  --initialization-actions="${INIT_ACTIONS}" \
```

Comment:
I did this as a stylistic choice because expanding them out in the command line resulted in a very long line.  I

> Since the init actions are all usable as-is with metadata substitution, most people will actually use the init actions we host in our own bucket.

I'm not sure I quite follow the point you're making, but if I understand correctly then you're suggesting I do away with the `INIT_BUCKET` variable in favor of just substituting it into lines 121-123.  I'm fine with this, but please confirm that this is what you meant.

Reply:
I meant to fully expand it in-line because it's generally easier to copy-paste a single command rather than a mini-script that has to set up state/variables first. However, the command line is long enough that it's fine to leave the snippet as is.